### PR TITLE
Added support of tempo variation (meta setTempo)

### PR DESCRIPTION
**Pull request by LenweSaralonde**

### CODE REVIEW

**Comment by LenweSaralonde** [comment]
Tests need to be adjusted since bwv-846.mid has tempo variation.

**Comment by tambien** [comment]
there's a lot of changes in here. is it possible to break it down into multiple PRs and validate each of them with tests?

File: src/Note.js [hunk 1]
```diff
@@ -11,8 +11,8 @@ class Note{
 		var note = new Note(json.midi, json.time, json.duration, json.velocity)
 		return note
 	}
-	
-	constructor(midi, time, duration=0, velocity=1){
+
+	constructor(midi, time, duration=0, velocity=1, channel, instrument){
```
Comment (leo6104) [inline src/Note.js:15]: Required variable should be defined before optional variables. 
`channel`, `instrument` variable should have default variable or change constructor arguments order.

```
constructor(midi, time, channel, instrument, duration=0, velocity=1) {
```

File: src/Track.js [hunk 2]
```diff
@@ -75,10 +81,12 @@ class Track {
 	 *                                  Pitch Notation like ('C#4')
 	 * @param  {Number} time     The time in seconds
 	 * @param  {Number} velocity The velocity value 0-1
+	 * @param  {Number} channel
+	 * @param  {Number} instrument
 	 * @return {Track} this
 	 */
-	noteOn(midi, time, velocity=1){
-		const note = new Note(midi, time, 0, velocity)
+	noteOn(midi, time, velocity=1, channel, instrument){
```
Comment (leo6104) [inline src/Track.js:89]: optional variable should be move to last.

**Review by leo6104 — CHANGES_REQUESTED** [review]
Almost LGTM. I also need to handle setTempo events in my project. 
Thanks. I will test it in my project and if have problems, i will report in here!

File: src/Note.js (hunk 1 above)
Comment (tambien) [reply src/Note.js:15]: thanks for the commends. tho i would prefer to include a default channel and instrument than switch the order of the API. both for usability and  backwards compatibility.

File: src/Midi.js [hunk 3]
```diff
@@ -81,35 +83,60 @@ class Midi {
...
 			const track = new Track()
-			track.id = i
-			this.tracks.push(track)
+			tracks.push(track)
 
+			let channel = -1
+			let instrument = -1
 			let absoluteTime = 0
 			trackData.forEach((event) => {
 				absoluteTime += Util.ticksToSeconds(event.deltaTime, this.header)
```
Comment (leo6104) [inline src/Midi.js:106]: `Util.ticksToSeconds` use `this.header.bpm` to convert `deltaTime`. Is it correct? 
I think it should be calculate seconds from current bpm (not header's bpm)

**Review by leo6104 — CHANGES_REQUESTED** [review]
I don't know exactly it is correct because i didn't fully understand midiconvert and midi format. but it seems ticksToSeconds should be calculate seconds with current bpm (not header.bpm)
What's your thought? @LenweSaralonde @tambien

**Comment by pr0jectile** [comment]
I'm also interested in this. Has anyone figured it out yet? :)

**Comment by tambien** [comment]
Thank you! but i'm closing this PR. big update to the lib and now has support for tempo changes

**Pull request closed without merging**